Update Ballerina Testing docs (#993)
ballerina-platform/ballerina-dev-website: swan-lake/learn/mocking.md

title: Mocking
description: Learn how to use Ballerina's built-in mocking API provided by the test to test modules
 independent from other modules and external endpoints.
keywords: ballerina, programming language, testing, mocking
permalink: /swan-lake/learn/mocking/
active: mocking
redirect_from:
  - /swan-lake/learn/mocking
  - /swan-lake/learn/mocking/
---

# Mocking

Mocking is useful to control the behavior of functions and objects to control the communication with other modules and external endpoints. A mock can be created by defining return values or replacing the entire object or function with a user-defined equivalent. This feature will help you to test the Ballerina code independently from other modules and external endpoints.

## Mocking Objects

The test module provides capabilities to mock an object for unit testing. This allows you to control the behavior of object member functions and values of member fields via stubbing or replacing the entire object with a user-defined equivalent. This feature will help you to test the Ballerina code independently from other modules and external endpoints.

Mocking of objects can be done in two ways.

> 1. Creating a test double (providing an equivalent object in place of the real)

Reviewer: ```suggestion
1. Creating double tests (providing an equivalent object in place of the real)
```

Author: "test double" is a jargon. So we can keep it as it is.

2. Stubbing the member function or member variable (specifying the behavior of functions and values of variables)

### Creating a test double

You can write a custom mock object and substitute it in place of the real object. The custom object should be made structurally equivalent to the real object via the mocking features in the test module.

***Example:***

> Let’s make changes to the example we tried out in the Getting Started section. In order to test the getRandomJoke function without actually calling the endpoint, we can use the test:mock function to mock the `get` remote function of the client object.

Reviewer: ```suggestion
Let’s make changes to the example in the Getting Started section. In order to test the` getRandomJoke` function without actually calling the endpoint, use the `test:mock` function to mock the `get` remote function of the client object.
```
Need to link to the example in the "Getting Started" section.

Author: Added the link and suggested changes in 99c63fb

Change the content of the ***main_test.bal*** file as follows:

```ballerina
// main_test.bal
import ballerina/test;
import ballerina/http;

public type MockHttpClient client object {
    	public remote function get(@untainted string path, public 
        	http:RequestMessage message = ()) returns 
        http:Response|http:ClientError {

        	http:Response res = new;
        	res.statusCode = 500;
        	return res;
    }
};

@test:Config {}
public function testTestDouble() {
  	clientEndpoint=<http:Client>test:mock(http:Client, new MockHttpClient());
  	string|error result = getRandomJoke("Sheldon");
  	test:assertTrue(result is error);
}
```

### Stubbing member functions and variables of an object

Instead of creating a test double you may also choose to create a default mock object and stub the functions to return a specific value or to do nothing.

#### Stubbing to return a specific value

***Example:***
The example in the Getting Started section shows how the get function of the client object can be stubbed to return a value. 

> Let’s make changes to the example we tried out in the Getting Started section.

Reviewer: ```suggestion
Let’s make changes to the example in the Getting Started section.
```
Need to link to the example.

### Mocking an imported function

The function specified with the **_@test:Mock {} _**annotation will be considered as a mock function that gets triggered every time the original function is called. The original function that will be mocked should be defined using the following annotation value fields.

*   ***moduleName*** : "&lt;moduleName&gt;" - Name of the module where the function to be mocked resides in. If the
 function is within the same module, this can be left blank or “.” (No module) can be passed. If the function is in a different module, but within the same project, just passing the module name will suffice. For functions in completely seperate modules, the fully qualified module name must be passed, which includes the orgName and the version. ie. orgName/module:version. For native function, the ballerina module needs to be specified.
*   ***functionName*** : "&lt;functionName&gt;" - Name of the function to be mocked.

> Mocking an imported function will apply the mocked function to every instance of the original function call. It is not limited to the test file where it is being mocked. 

Reviewer: ```suggestion
Mocking an imported function will apply the mocked function to every instance of the original function call. It is not limited to the test the file, which is being mocked. 
```

 every time the original function is called. Subsequent to the declaration, the function call should be stubbed using the available function mocking features. Different behaviors can be defined for different test cases if required.

***Example:***

**main.bal**
```ballerina
// main.bal
import ballerina/io;

public function addValues(int a, int b) returns int {
	return intAdd(a, b);
}

public function intAdd(int a, int b) returns int {
	return (a + b);
}
```
**main_test.bal**
```ballerina
// main_test.bal
import ballerina/test;

@test:MockFn { functionName: "intAdd" }
test:MockFunction intAddMockFn = new();

@test:Config {}
function testReturn() {
	test:when(intAddMockFn).thenReturn(20);
	test:when(intAddMockFn).withArguments(0, 0).thenReturn(-1);

	test:assertEquals(intAdd(10, 6), 20, msg = "function mocking failed");
	test:assertEquals(intAdd(0, 0), -1,
    	msg = "function mocking with arguments failed");
}

@test:Config {}
function testCall() {
	test:when(intAddMockFn).call("mockIntAdd");
	test:assertEquals(addValues(11, 6), 5, msg = "function mocking failed");
}

public function mockIntAdd(int a, int b) returns int {
	return (a - b);
}
```

## What's Next
 

> Now that you are aware of the details on writing tests, learn the different options that can be used when [Executing

Reviewer: ```suggestion
Now, that you are aware of the details on writing tests, learn the different options that can be used when [Executing
```
